### PR TITLE
Fused backward pass kernel

### PR DESCRIPTION
**Pull request by unixpickle**
##### Description

This implements the backward pass as a single CUDA kernel call. This should simultaneously reduce kernel call overhead and memory I/O.

##### Benchmark

I benchmarked the time of a forward+backward operation with shape `8 x 512 x 65536` on my Titan X.
 * Old time: **67.4ms**
 * New time: **33.1ms**
 * Speedup: **~50%**

I did not bother to separate out the cost of the forward (which should be unchanged), since a forward+backward best reflects the actual calls during training of a model.

Benchmarked with this in a notebook:

```python
from accelerated_scan.warp import scan as scan_warp
from accelerated_scan.ref import scan as scan_ref
import torch

x = torch.nn.Parameter(torch.randn(8, 512, 65536).cuda()) * 0.2
y = torch.nn.Parameter(torch.randn(8, 512, 65536).cuda())
grad_out = torch.randn_like(x)

def fn():
    warp_out = scan_warp(x, y)
    warp_grad = torch.autograd.grad(warp_out, (x, y), grad_out)
    sum(x.sum().item() for x in warp_grad)

%timeit fn()
```

##### Testing

I tested against the reference implementation as follows. However, I have not run the full test suite.

```python
from accelerated_scan.warp import scan as scan_warp
from accelerated_scan.ref import scan as scan_ref
import torch

x = torch.nn.Parameter(torch.randn(8, 32, 1024).cuda()) * 0.2
y = torch.nn.Parameter(torch.randn(8, 32, 1024).cuda())
ref_out = scan_ref(x, y)
warp_out = scan_warp(x, y)
grad_out = torch.randn_like(ref_out)

ref_grad = torch.autograd.grad(ref_out, (x, y), grad_out)
warp_grad = torch.autograd.grad(warp_out, (x, y), grad_out)
for i, (actual, expected) in enumerate(zip(ref_grad, warp_grad)):
    assert torch.allclose(actual, expected, atol=1e-4), f'{i=} {actual=} {expected=}'
print('PASS')
```

### CODE REVIEW

**Comment by proger** [comment]
Thank you @unixpickle!

Avoiding tensor allocations for gate padding and subsequent postprocessing saves quite a bit of time. I'm going to add backward tests to `tests/bench.py`.

Tuple packing from #6 cleanly interacts with shifted loading. The static offset template parameter enables simultaneous slicing from the left and padding from the right with positive values as needed for gate padding, and reversed with negative for output.

Will leave some questions in the comments.

I am wondering if recomputing forward activations in the backward pass will save even more time — this way we keep only one entry point in the cuda kernel file.

File: accelerated_scan/warp.cuh [hunk 1]
```diff
@@ -342,4 +344,371 @@ warpscan_forward(const at::Tensor &gates, const at::Tensor &tokens, const at::Te
...
+    using weight_t = typename Tuple::Type;
+
+    const weight_t* rawPtr = reinterpret_cast<const weight_t *>(ptr);
+    Tuple x;
+    for (int i = 0; i < Tuple::Size; i++) {
+        const int idx = index * Tuple::Size + i + offset;
+        if (idx >= minIdx * Tuple::Size && idx < maxIdx * Tuple::Size) {
+            x.data[i] = rawPtr[idx];
+        } else {
+            x.data[i] = 0.0;
```
Comment (proger) [inline accelerated_scan/warp.cuh]: When offset is 1 we need to pad with values of 1 — I'm a bit confused why it doesn't show up on your test. Would like to play with this.

```python
padded_shifted_gates = torch.cat([gates, torch.ones_like(gates[:, :, :1])], dim=-1)[:, :, 1:].contiguous()
d_states = scan_forward(padded_shifted_gates, grad_output, reverse=True)
padded_outputs = torch.cat([torch.zeros_like(states[:, :, :1]), states], dim=-1)[:, :, :-1]
```

**Comment by proger** [comment]
[Benchmark](https://github.com/proger/accelerated-scan/commit/5cb9403b6ec2d39a86c9c8b5da77ee8b3587ddea) shows improvements from 33% on short sequences to 43% on longer on a 3090.

fused

File: accelerated_scan/warp.cuh (hunk 1 above)
Comment (unixpickle) [reply accelerated_scan/warp.cuh]: I think it's because the first gate in a scan is unused (since the -1'th acc is 0). So it won't matter what we pad.

**Comment by unixpickle** [comment]
> I am wondering if recomputing forward activations in the backward pass will save even more time — this way we keep only one entry point in the cuda kernel file.

I'm pretty sure this will use the same amount of memory for autograd (cached activations) while using more compute.

I'd be interested in trying to unify kernel entrypoints though. Probably just some extra arguments are enough to make it work. I was thinking we could also unify the logic in the C++ functions that select various block sizes. We could potentially use fancy C++ templates for that, but haven't thought about it enough.

**Comment by proger** [comment]
Love the idea of unifying entrypoints. Even C macros would do here I believe.

**Comment by unixpickle** [comment]
Merged the two codepaths a lot more. There are still two entrypoint functions, but they are quite small and both call the same underlying function.

**Comment by proger** [comment]
Thank you! Pulling this and making a release.
